### DISCUSSION TITLE
How to authenticate to install packages from Github Packages?

### DISCUSSION THREAD

**Discussion by marccals** [community/community#25282] [Actions]
<p>Hello All,</p>
<p>We’re migrating our build system to Github. Some of our repositories build nugets which are consumed by many other repositories in Github. To store these nugets we will use Github Packages Registry</p>
<p>For the repositories which need to install nugets from Github Packages, we’ve been investigating which is the best way to authenticate from Github actions to Github package registry to have rights to install all needed packages . In your documentation have find the following:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/a/4/a4fe2c2b2bff2702a3acfd6fa598f6c100f7daa3.png" data-download-href="/uploads/short-url/nxARLSQQ0yi9Dyxpc4Jhzm753fJ.png?dl=1" title="2022-02-21_09h14_04" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/1679416/181090063-d065151d-a371-4aed-9925-98af9fbd1387.png" alt="2022-02-21_09h14_04" data-base62-sha1="nxARLSQQ0yi9Dyxpc4Jhzm753fJ" width="690" height="328" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/a/4/a4fe2c2b2bff2702a3acfd6fa598f6c100f7daa3_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use href="#far-image"></use></svg><span class="filename">2022-02-21_09h14_04</span><span class="informations">756×360 20.8 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use href="#discourse-expand"></use></svg>
</div></a></div>
<p>According to your documentation the only way to install packages associated to other private repository it’s using a personal access token (PAT). In our automatic builds in Github we don’t want to use a PAT stored as a secret. A PAT it’s associated to a personal account and If this account is removed, someone deletes/expires accidentally the PAT, etc… all our builds will break.</p>
<p>Having the step to install nuget packages from other repositories associated to a PAT we think it has a lot of drawbacks.</p>
<p>One approach we tried, but it didn’t work was to generate a token from a Github app with read permissions on Github package with this action <a href="https://github.com/tibdex/github-app-token" class="inline-onebox" rel="noopener nofollow ugc">GitHub - tibdex/github-app-token: Impersonate a GitHub App in a GitHub Action</a>, when registering the nuget sources</p>
<p><code>nuget sources Add -Name MyNugetSource -Source "$env:GPR_URL" -UserName "fakeUser" -Password "$env:GPR_Token_from_Github_APP"</code></p>
<p>But we got authentication error, when looking for the packages to install during the build.</p>
<p>Is there a better way to authenticate to be able to install our nuget packages from other repositories without using PAT?</p>
<p>Thanks</p>

**Comment by mattwelke** [comment, marked as answer]
<p>As far as I know, the limitation described in the documentation still exists. Workflows for a repo in an organization cannot read packages in other repos even if those repos are part of the same organization. That would make sense as something an enterprise would want to do, but GitHub doesn’t have it yet.</p>
<p>Because you always need a PAT with the read:packages scope to download packages from the registry (even if the repo associated with the package is a public repo - that’s another limitation right now), I recommend working around this by creating a GitHub account dedicated to CI/CD use cases. That account could be managed by one or more people in the organization and its token lifecycles would thankfully be separate from tokens associated with personal GitHub accounts.</p>

**Reply by rowanfreeman-acutro** [reply]
This seems like a weird oversight. I can publish my NuGet package easily enough, but then a different repository within the same organisation can't use it.

All the documentation says to use a personal access token. But that's really bad practice to use a PAT from an individual account within an organisation.

Question for GitHub: are there any plans to address this?

**EDIT**

Ok, I think I figured this out.

This worked for me. I have a .NET application that references a NuGet package that is hosted in the same organisation but different repository.

Go to your org's **package settings**. Should be `https://github.com/orgs/<Organisation>/packages/nuget/<PackageId>/settings`.
Then under **Manage Actions access**, add the repository(ies) that need access to the package. Read access is fine.

For your application, make sure it has a `nuget.config` file at the root.

```xml
<?xml version="1.0" encoding="utf-8"?>
<configuration>
    <packageSources>
        <add key="github" value="https://nuget.pkg.github.com/<Organisation>/index.json"/>
    </packageSources>
</configuration>
```

Then, for your application's workflow, use the `NuGetPackageSourceCredentials_<key>` environment variable.

Example:

```yml
jobs:
  build:
    name: Build and test
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v4

      - name: Setup .NET
        uses: actions/setup-dotnet@v4
        with:
          dotnet-version: 8.0.x

      - name: Build
        run: dotnet build --configuration Release
        env:
          NuGetPackageSourceCredentials_github: Username=${{ github.actor }};Password=${{ secrets.GITHUB_TOKEN }}

      - name: Test
        run: dotnet test --configuration Release --no-build
```
